### PR TITLE
Feature/fuller creds

### PR DESCRIPTION
**Pull request by berkes**
This adds additional fields to the credential output.

We start simple with validUntil and validFrom datefields- which are optional both. validFrom is always set in practice. validUntil not.

To achieve this, we've refactored the JSON generation first. Since building the json structure with conditional fields turns messy really fast, we've opted for a serializer-pattern. As follows:

image

So, we have "models" as POPOs (plain old python objects), in a tree structure, with `EduCredential` as root, that we can serialize with corresponding serializers. This tree of "models" can be instantiated from actual Django Record BadgeInstance model.

Basically: we can now serialize Django BadgeInstance records as an OB3 JSON structure.

PR is in draft because it continues on #155

### CODE REVIEW

File: apps/ob3/serializers.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+            source='participation',
+            required=False,
+            allow_null=True,
+    )
+    alignment = AlignmentSerializer(many=True)
+
+    def to_representation(self, instance):
+        ret = super().to_representation(instance)
+        """Convert the id to a URL"""
+        ret['id'] = f"https://example.com/assertions/{ret['id']}"
```
Comment (Iso5786) [inline apps/ob3/serializers.py:126]: Update to edubadges nl URL? Such as https://demo.edubadges.nl/public/assertions/l0FxH-o2QZ2p4CyDfmOdBw

File: apps/ob3/models.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+        if "extensions:LanguageExtension" in badge_class.extension_items:
+            in_language = badge_class.extension_items["extensions:LanguageExtension"]["Language"]
+
+        if "extensions:ECTSExtension" in badge_class.extension_items:
+            ects = badge_class.extension_items["extensions:ECTSExtension"]["ECTS"]
+
+        if "extensions:EducationProgramIdentifierExtension" in badge_class.extension_items:
+            education_program_identifier = badge_class.extension_items["extensions:EducationProgramIdentifierExtension"]["EducationProgramIdentifier"]
+
```
[the remark this answers is not in the document]
Comment (berkes) [reply apps/ob3/models.py:76]: I wanted to make this a polymorphic (sorry for that word, I hate it ;) serializer, with e.g. an Abstract Class Pattern. So `EducationProgramIdentifierExtensionSerializer`, `LanguageExtension` etc. And the python logic to select the correct serializer class and instantiate it in 

I don't know Python, nor DRF, well enough to know if it offers such generic subclass instantiation, or what a Pythonic way of doing this. Therefore I have postponed this :)

**Comment by berkes** [comment]
To discuss:

- [ ] Identity Object
- [ ] ID urls: where to point to, how to deal with demo.edubadges.nl ?
- [ ] Generalization for Serialized extensions?

File: apps/ob3/serializers.py (hunk 1 above)
Comment (berkes) [reply apps/ob3/serializers.py]: Maar dan linken naar de ENV var versie.

File: apps/ob3/serializers.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+            child=serializers.CharField(),
+            read_only=True,
+            default=["Profile"]
+        )
+    name = serializers.CharField()
+
+    def to_representation(self, instance):
+        ret = super().to_representation(instance)
+        """Convert the id to a URL"""
+        ret['id'] = f"https://example.com/issuers/{ret['id']}"
```
[the remark this answers is not in the document]
Comment (berkes) [reply apps/ob3/serializers.py]: https://demo.edubadges.nl/public/institutions/9-TYGj4dTn-nW8CGfwGjAQ bijv

File: apps/ob3/models.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+class Credential:
+    def __init__(self, issuer, valid_from, credential_subject, **kwargs):
+        self.issuer = issuer
+        self.valid_from = valid_from
+        self.credential_subject = credential_subject
+
+        self.valid_until = kwargs.get('valid_until', None)
+
+class Achievement:
+    FIELDS = [
```
[the remark this answers is not in the document]
Comment (berkes) [reply apps/ob3/models.py:24]: Implemented in another PR at #160 160

[the remark this answers is not in the document]
Comment (Iso5786) [reply apps/ob3/models.py:119]: Looks good

File: apps/ob3/models.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Iso5786) [reply apps/ob3/models.py:147]: Fine with me
